### PR TITLE
DEV/MNT: migrate to GitHub Actions and pyproject.toml [LCLSPC-603]

### PR DESCRIPTION
**Pull request by klauer**
## Description
* Upgrading by way of https://github.com/pcdshub/pcds-python-migration-tools
* hutch-python no longer uses Travis CI and has migrated to GitHub Actions for continuous integration, testing, and documentation deployment.
* hutch-python has been migrated to use setuptools-scm, replacing versioneer, as its version-string management tool of choice.
* hutch-python has been migrated to use the modern ``pyproject.toml``, replacing ``setup.py`` and related files.
* Older language features and syntax found in the repository have been updated to Python 3.9+ standards by way of ``pyupgrade``.
* Sphinx 6.0 is now supported for documentation building. ``docs-versions-menu`` replaces ``doctr-versions-menu`` and ``doctr`` usage for documentation deployment on GitHub Actions.  The deployment key is now no longer required.
* `pyupgrade` was used to update annotations.
* `pycln` was used to clean up imports

## Motivation and Context
* Migrating because we want to migrate our projects from Travis CI for one
* I also copy/pasted this text from another repository and linked an unrelated issue. Oops.

## How Has This Been Tested?
* GitHub Actions

## Where Has This Been Documented?
This PR and Jira [LCLSPC-603](https://jira.slac.stanford.edu/browse/LCLSPC-603)

### CODE REVIEW

**Comment by klauer** [comment]
One test consistently failing: `FAILED hutch_python/tests/test_tstpython.py::test_tstpython_ipython`
Need to look into details...

Some other weird output in the log, but discussed in side channels it's probably intentional-ish

**Comment by tangkong** [comment]
This might be the same test that's been failing? https://github.com/pcdshub/hutch-python/issues/354  (yep https://app.travis-ci.com/github/pcdshub/hutch-python/jobs/588574480#L1011)

This also probably closes #335 ?

**Review by ZLLentz — APPROVED** [review]
We can fix that test in a differently scoped PR

File: .coveragerc [hunk 1]
```diff
@@ -9,4 +9,3 @@ omit =
     #tests
     *test*
     hutch_python/tests/*
-    hutch_python/cookiecutter/*
```
[the remark this answers is not in the document]
Comment (klauer) [reply .coveragerc]: Good catch; fixed in 6a709a6

**Review by tangkong — APPROVED** [review]
👍 All of these migration PRs are starting to blur together

**Comment by klauer** [comment]
(Merging and bypassing branch restrictions; reviewers indicated that IPython-related bug should be solved in a separate PR)

**Pull request merged**